mem: Stride Prefetcher Fix (#1448)

Setu-Gupta opened the pull request:
This PR fixes the issues mentioned in #1448.

**Note that this contribution is the result of a joint collaboration with @AbhishekUoR**

This PR introduces the following 4 changes:
1. It changes the addresses which are used to compute the stride to cache line aligned addresses (the current version uses word aligned addresses)
2. It correctly returns if the stride does not match (as opposed to issuing prefetches using the new stride incorrectly)
3. It returns if the new stride is 0, indicating multiple reads from the same cache line.
4. It removes code which is no longer necessary after the addition of changes number 1 and 3.

Change-Id: Ic346d0e15df6d07e2b93289c8d6b89b4c2f45a34

Setu-Gupta committed:
mem: Fixed bugs in the Stride Prefetcher

On branch stride_fix
Changes to be committed:
	modified:   src/mem/cache/prefetch/stride.cc

Change-Id: I4a3f4959ffd4ad526609103dfde0e3ecec82ff34
Co-authored-by: Abhishek Shailendra Singh <abs218@leigh.edu>

Setu-Gupta committed:
mem: Removed redundant code

On branch stride_fix
Your branch is up to date with 'origin/stride_fix'.

Changes to be committed:
	modified:   src/mem/cache/prefetch/stride.cc

Change-Id: Ic346d0e15df6d07e2b93289c8d6b89b4c2f45a34
Co-authored-by: Abhishek Shailendra Singh <abs218@leigh.edu>

odanrc commented on src/mem/cache/prefetch/stride.cc:
```diff
@@ -149,6 +149,11 @@ Stride::calculatePrefetch(const PrefetchInfo &pfi,
         // Hit in table
         int new_stride = pf_addr - entry->lastAddr;
+
+        // Do nothing on repeated memory access
```
Why is this the correct approach? My initial thought is that deciding if 0-strides should not modify the confidence should at the very least be a configuration parameter.

If we conclude that 0 strides should not update the confidence the code in line 157 needs to be updated.

Setu-Gupta replied:
I agree. I've modified the code and now the users can choose between the granularity of cache line or word address. Only if the granularity is set to cache lines, we will ignore zero strides

odanrc replied:
Alright, I agree with/understand the change to line 162; however, I don´t understand why the zero-stride check only applies to the cache-line scenario. Could you clarify?

Setu-Gupta replied:
The reason why I was treating the the zero stride case separately was because I was trying to guard against multiple accesses to the same cache line. This can only happen if the prefetcher is operating at the granularity of the cache lines. If it operating at the granularity of word addresses, then we don't care even if we observe zero strides

odanrc commented on src/mem/cache/prefetch/stride.cc:
```diff
@@ -169,17 +174,15 @@ Stride::calculatePrefetch(const PrefetchInfo &pfi,
 
         entry->lastAddr = pf_addr;
 
+        // Return if we don't see a stride match
```
I think this would be a bug. The point of having the confidence threshold is exactly to decide when the predicted stride is relevant or not. If we have 99% confidence that an access to X will have a stride S, but this time we got a stride of S2, then the confidence indicates that S2 was just an outlier, and that we will probably see S again in the next access.

Setu-Gupta replied:
I agree. I've changed the code to reflect this

giactra commented:
@rjc-arch

Setu-Gupta committed:
mem: Added a paramter to the stride prefetcher

The users can use this parameter to choose the granularity of the the
prefetcher

Changes to be committed:
	modified:   src/mem/cache/prefetch/Prefetcher.py
	modified:   src/mem/cache/prefetch/stride.cc
	modified:   src/mem/cache/prefetch/stride.hh

Change-Id: Id6c6a5b901008783669b5bf083b37b4952011ae4
Co-authored-by: Abhishek Shailendra Singh <abs218@leigh.edu>

odanrc commented on src/mem/cache/prefetch/stride.cc:
```diff
@@ -174,9 +181,9 @@ Stride::calculatePrefetch(const PrefetchInfo &pfi,
+        // Round strides up to at least 1 cacheline
+        int prefetch_stride = entry->stride;
+        if (abs(prefetch_stride) < blkSize) {
             prefetch_stride = (new_stride < 0) ? -blkSize : blkSize;
```
I think new_stride should have been replaced by prefetch_stride here too

Setu-Gupta replied:
Yes, you are correct. Thanks for catching this error. I've fixed this

odanrc commented on src/mem/cache/prefetch/stride.cc:
```diff
@@ -133,7 +134,8 @@ Stride::calculatePrefetch(const PrefetchInfo &pfi,
 
     // Get required packet info
-    Addr pf_addr = pfi.getAddr();
+    Addr pf_addr = useCachelineAddr ? blockAddress(pfi.getAddr())
```
Is there a performance benefit on not using block addresses given that we end up rounding the prefetch strides to block addresses anyway?

Setu-Gupta replied:
In a micro-benchmark that I wrote I saw a very small improvement in IPC when I was operating at the cache line granularity. This mainly happens because when we operate on word granularity, we may issue multiple prefetches for the same cache line for different words. Although these prefetches can be coalesced into one at the prefetch queue, we would still be exhausting the bandwidth of the prefetcher if we issue a lot of redundant prefetches. That's the reason why I've set the default parameters of the prefetcher such that it operates on cache line granularity

Setu-Gupta committed:
mem: Fixed a bug in Stride Prefetcher

Change-Id: Ib13596723d536323f172cf89553608ce2fea7ffc
Co-authored-by: Abhishek Shailendra Singh <abs218@leigh.edu>

odanrc reviewed (approved):
The changes seem reasonable, but since I imagine the stride prefetcher is one of the most used prefetchers by gem5 users, and these changes effectively change its behavior, we will probably need to add an entry to the release notes @BobbyRBruce .

BobbyRBruce commented:
> The changes seem reasonable, but since I imagine the stride prefetcher is one of the most used prefetchers by gem5 users, and these changes effectively change its behavior, we will probably need to add an entry to the release notes @BobbyRBruce .

Added here: #1460
